toolkit/tools/precacher/precacher.go

```diff
@@ -0,0 +1,409 @@
+	buildDir          = app.Flag("worker-dir", "Directory to store chroot while running repo query.").String()
+
+	concurrentNetOps = app.Flag("concurrent-net-ops", "Number of concurrent network operations to perform.").Default(defaultNetOpsCount).Uint()
+)
+
+func main() {
+	app.Version(exe.ToolkitVersion)
+	kingpin.MustParse(app.Parse(os.Args[1:]))
+	logger.InitBestEffort(*logFile, *logLevel)
+
+	prof, err := profile.StartProfiling(profFlags)
+	if err != nil {
+		logger.Log.Warnf("Could not start profiling: %s", err)
+	}
+	defer prof.StopProfiler()
+
+	timestamp.BeginTiming("precacher", *timestampFile)
+	defer timestamp.CompleteTiming()
+
+	// Get the state of the current repo from a snapshot
+	rpmSnapshot, err := rpmSnapshotFromFile(*snapshot)
+	if err != nil {
+		logger.PanicOnError(err)
+	}
+
+	// See what is available online from the upstream repos
+	availablePackages, err := getAllRepoData(*repoUrls, *workerTar, *buildDir)
+	if err != nil {
+		logger.PanicOnError(err)
+	}
+
+	logger.Log.Infof("Pre-caching: Found %d available packages", len(availablePackages))
+	if logger.Log.IsLevelEnabled(logrus.DebugLevel) {
+		for _, pkg := range availablePackages {
+			logger.Log.Debugf("Found package: %s", pkg)
+		}
+	}
+
+	// For each package in the snapshot, check if it is available online and try to download it
+	downloadedPackages, err := downloadMissingPackages(rpmSnapshot, availablePackages, *outDir, *concurrentNetOps)
+	if err != nil {
+		logger.PanicOnError(err)
+	}
+
+	logger.Log.Infof("Pre-caching: Downloaded %d packages into the cache", len(downloadedPackages))
+	err = writeSummaryFile(*outputSummaryFile, downloadedPackages)
+	if err != nil {
+		logger.PanicOnError(err)
+	}
+}
+
+func rpmSnapshotFromFile(snapshotFile string) (rpmSnapshot *repocloner.RepoContents, err error) {
+	err = jsonutils.ReadJSONFile(snapshotFile, &rpmSnapshot)
+	return
+}
+
+// getAllRepoData returns a map of package names to URLs for all packages available in the given repos. It uses
+// a chroot to run repoquery.
+func getAllRepoData(repoUrls []string, workerTar, buildDir string) (namesToUrls map[string]string, err error) {
+	const (
+		leaveChrootOnDisk = false
+	)
+	timestamp.StartEvent("pull available package data from repos", nil)
+	defer timestamp.StopEvent(nil)
+
+	// Create a chroot to run repoquery in
+	// Create the directory 'buildDir' if it does not exist
+	exists, err := file.DirExists(buildDir)
+	if err != nil {
+		err = fmt.Errorf("failed to check if directory %s exists: %w", buildDir, err)
+		return nil, err
+	}
+	if !exists {
+		logger.Log.Infof("Pre-caching: Creating 1st time chroot directory %s", buildDir)
+		err = os.MkdirAll(buildDir, 0755)
+		if err != nil {
+			err = fmt.Errorf("failed to create directory %s: %w", buildDir, err)
+			return nil, err
+		}
+	}
+	queryChroot, err := createChroot(workerTar, buildDir, leaveChrootOnDisk)
+	if err != nil {
+		err = fmt.Errorf("failed to create chroot: %w", err)
```

Comment:
Did you mean "newlines" instead of "spaces"?

Reply:
Ah, copy pasted this from elsewhere in the code that predates that switch. I'll update.
(Although, I'm leaning towards mabybe doing `"error msg:\n\t%w"` or something, it can be hard to follow the errors on the console sometimes since they now lack the formatted red header on every line.

Reply:
Ah, looks like I went back to the old style elsewhere as well. Fixed those too.